src/Cli/dotnet/commands/dotnet-workload/install/FileBasedInstaller.cs

```diff
@@ -284,87 +295,162 @@ public IEnumerable<WorkloadDownload> GetDownloads(IEnumerable<WorkloadId> worklo
+                             !featureBandsWithWorkloadInstallationRecords.Contains(referencingFeatureBand))
                     {
-                        File.Delete(unneededPackRecord);
+                        //  Remove pack record if corresponding feature band is not installed, or does not have any workloads installed
+                        featureBandsToRemove.Add(referencingFeatureBand);
                     }
-
-                    // If there are no pack records left in the directory we garbage collected, we need to delete the pack and its director(ies).
-                    if (!bandPackRecordPaths.Except(unneededBandRecordPaths).Any())
+                    else if (referencingFeatureBand.Equals(_sdkFeatureBand) && !garbageCollector.PacksToKeep.Contains((packId, packVersion)))
                     {
-                        Directory.Delete(packVersionDir);
-                        var deletablePack = GetPackInfo(packVersionDir);
-                        if (deletablePack == null)
-                        {
-                            // Pack no longer exists in manifests, get pack info from installation record
-                            deletablePack = JsonSerializer.Deserialize(jsonPackInfo, typeof(PackInfo)) as PackInfo;
-                        }
-                        DeletePack(deletablePack);
+                        //  For current feature band, garbage collect packs that the garbage collector didn't mark as ones to keep
+                        featureBandsToRemove.Add(referencingFeatureBand);
                     }
                 }
 
-                if (!Directory.GetFileSystemEntries(packIdDir).Any())
+                if (!featureBandsToRemove.Any())
+                {
+                    continue;
+                }
+
+                // Save the pack info in case we need to delete the pack (the pack type is needed to know the path to the pack and how to delete it)
+                var jsonPackInfo = File.ReadAllText(GetPackInstallRecordPath(packId, packVersion, featureBandsToRemove.First()));
+
+                foreach (var featureBand in featureBandsToRemove)
                 {
-                    Directory.Delete(packIdDir);
+                    File.Delete(GetPackInstallRecordPath(packId, packVersion, featureBand));
+                }
+
+                var installationRecordDirectory = Path.GetDirectoryName(GetPackInstallRecordPath(packId, packVersion, featureBandsToRemove.First()));
+                if (!Directory.GetFileSystemEntries(installationRecordDirectory).Any())
+                {
+                    //  There are no installation records for the workload pack anymore, so we can delete the pack
+                    var packToDelete = JsonSerializer.Deserialize<PackInfo>(jsonPackInfo);
+                    DeletePack(packToDelete);
+
+                    //  Delete now-empty pack installation record directory
```

Comment:
Should the comment be `non-empty` instead of `now-empty`?

Reply:
No, the directory is now empty after deleting the pack installation records (which is what the if statement above checks).